### PR TITLE
Detection and handling of autosaved session at start-up (Issue 160)

### PR DESCRIPTION
**Pull request by tparchambault**
Still some minor refactoring, corner cases outstanding. I wanted to subject the branch to the CI pipeline.

### CODE REVIEW

File: python/fapolicy_analyzer/tests/test_state_manager.py [hunk 1]
```diff
@@ -53,11 +56,63 @@ def populated_changeset_list():
...
+    "/home/toma/Development/CTC/data_space/man_from_mars.txt": "Add",
+    "/home/toma/Integration.json": "Add"
+}''')
+    fp1.write('''{
+    "/home/toma/Development/CTC/data_space/man_from_mars.txt": "Add",
+    "/home/toma/Development/CTC/data_space/this/is/a/longer/path/now_is_the_time.txt": "Del",
+    "/home/toma/Integration.json": "Add"
+}''')
+    fp0.close()
+    fp1.close()
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/tests/test_state_manager.py:76]: Good point about the context manager, 'with'. The contents of those files actually doesn't matter, as long as the exact data finds its way into the internal data structures. I'll set it to something less toma and/or CTC specific nonetheless.

File: python/fapolicy_analyzer/ui/main_window.py [hunk 2]
```diff
@@ -142,6 +146,27 @@ def on_start(self, *args):
...
+            # Raise the modal  "Prior Session Detected" dialog to
+            # prompt the user to immediate restore the prior edit session
+            response = self.__AutosaveRestoreDialog()
+
+            if response == Gtk.ResponseType.YES:
+                try:
+                    stateManager.restore_previous_session()
+                    self.get_object("restoreMenu").set_sensitive(False)
+                except Exception:
+                    print("Restore failed")
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/main_window.py:166]: That would be a good thing. Will do that.

File: python/fapolicy_analyzer/tests/test_state_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/tests/test_state_manager.py:76]: The important things is actually selecting the right file based on the embedded timestamp in the filename.

File: python/fapolicy_analyzer/ui/main_window.py [hunk 3]
```diff
@@ -156,3 +181,43 @@ def on_changeset_updated(self):
...
+
+    Yes: Immediately loads your prior session
+
+    No: Continue starting fapolicy-analyzer.
+
+        Your prior session will still be available
+        and can be loaded at any point during
+        this current session by invoking 'Restore'
+        under the 'File' menu.
+
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/main_window.py:215]: Ok.

File: python/fapolicy_analyzer/ui/state_manager.py [hunk 4]
```diff
@@ -229,6 +214,51 @@ def __force_cleanup_autosave_sessions(self):
...
+                print("Attempting to restore session file: {}".format(f))
+                self.open_edit_session(f)
+                print("Returned from open_edit_session({})".format(f))
+                self.__cleanup_autosave_sessions()
+                bReturn = True
+                print("SUCCESS")
+                break
+
+            except Exception:
+                print("FAIL: Restoring {} load failure".format(f))
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/state_manager.py:257]: Maybe. It depends on where in the restore process. The code iterates through all available autosaved files from newest to oldest, stopping at the first successful load. As long as the failures propagate to the main_window level, then we'll display the toaster pop-up messaging.

File: python/fapolicy_analyzer/tests/test_state_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/tests/test_state_manager.py:76]: Changes made; Passes the unit-test suite run locally.

File: python/fapolicy_analyzer/ui/__main__.py [hunk 5]
```diff
@@ -9,14 +10,39 @@
...
 
+def XdgDirPrefix(strBaseName):
+    """Prefixes the file basename, strBaseName, with the XDG_STATE_HOME
+    directory, creates the directory if needed, and verifies that it is writable
+    by the effective user
+    """
+    # Use the XDG_STATE_HOME env var, or $(HOME)/.local/state/
+    _home = os.path.expanduser('~')
+    xdg_state_home = os.environ.get('XDG_STATE_HOME') or \
+        os.path.join(_home, '.local', 'state')
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/__main__.py:21]: Changed as suggested.

File: python/fapolicy_analyzer/ui/main_window.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/main_window.py:166]: stateManager.restore_previous_session() will return False if the autosaved files are corrupted. The toaster pop-up made conditional based on that function's return value.

File: python/fapolicy_analyzer/ui/main_window.py [hunk 6]
```diff
@@ -95,7 +95,11 @@ def on_openMenu_activate(self, menuitem, data=None):
 
     def on_restoreMenu_activate(self, menuitem, data=None):
         logging.debug("Callback entered: MainWindow::on_restoreMenu_activate()")
-        pass
+        try:
+            stateManager.restore_previous_session()
+            self.get_object("restoreMenu").set_sensitive(False)
+        except Exception:
+            print("Restore failed")
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/main_window.py:102]: Ditto as above.

File: python/fapolicy_analyzer/ui/main_window.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/main_window.py:166]: Added code at the highest UI level (MainWindow) to present a pop-up notification on failure.

File: python/fapolicy_analyzer/ui/state_manager.py [hunk 7]
```diff
@@ -187,22 +166,28 @@ def save_edit_session(self, strJsonFile):
...
-            logging.debug("Loaded dict = ", d)
-            listPA = self.__path_action_dict_to_list(d)
-            logging.debug("StateManager::open_edit_session():{}".format(listPA))
+            try:
+                d = json.load(fp, object_pairs_hook=OrderedDict)
+                logging.debug("Loaded dict = ", d)
+                listPA = self.__path_action_dict_to_list(d)
+                logging.debug("StateManager::open_edit_session():{}".format(listPA))
+            except Exception as e:
+                print(e, "json.load() failure")
```
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/state_manager.py:178]: As above: Added code at the highest UI level (MainWindow) to present a pop-up notification on failure.

File: python/fapolicy_analyzer/ui/main_window.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tparchambault) [reply python/fapolicy_analyzer/ui/main_window.py]: Done. Added to strings.py and rebuilt pot file.
